File: appAcademyOpen/JsPractice/anagrams.js
// Write a function, `anagrams(str1, str2)`, that takes in two words and returns 
// a boolean indicating whether or not the words are anagrams. Anagrams are 
// words that contain the same characters but not necessarily in the same order. 
// Solve this without using Array.prototype.sort.
// 
// Examples:
// anagrams('listen', 'silent') => true
// anagrams('listen', 'potato') => false

function anagrams(str1, str2){
    let counter = {};

    for(let i = 0; i < str1.length; i++){
        if(str1[i] in counter){
            counter[str1[i]] ++;
        }else{
            counter[str1[i]] = 1;
        }
    }

    for(let j = 0; j < str1.length; j++){
        if(str2[j] in counter){
            counter[str2[j]] --;
        }
    }
    console.log(counter)
    for(let keys in counter){
        if(counter[keys] !== 0) return false;
    }

    return true;
}

console.log(anagrams('listen', 'silent')) //=> true
console.log(anagrams('listen', 'potato') )//=> false

// time complexity:  O(n)
// space coplexity: O(n)

File: appAcademyOpen/Ruby Practice/strings/anagrams.rb
# Write a function `anagrams(str1, str2)` that takes in two words and returns a 
# boolean indicating whether or not the words are anagrams. Anagrams are words 
# that contain the same characters but not necessarily in the same order. Solve 
# this without using `Array#sort` or `Array#sort_by`.

# n = input str1 length
# m = input str2 length
# n === m 

# time coplexity: O(n)
def anagrams(str1, str2)
  counter = Hash.new(0)
  str1.each_char do |char|
    counter[char] += 1
  end

  str2.each_char do |char|
    counter[char] -= 1
  end

  counter.each_value do |v|
    return false if v != 0 
  end
  return true
end